jina/peapods/peas/__init__.py

```diff
@@ -1,338 +1,174 @@
-        if 'role' in self.args and self.args.role == PeaRoleType.PARALLEL:
-            self.name = f'{self.name}-{self.args.pea_id}'
-        if 'log_id' in self.args and 'log_config' in self.args:
-            self.logger = JinaLogger(self.name,
-                                     log_id=self.args.log_id,
-                                     log_config=self.args.log_config)
-        else:
-            self.logger = JinaLogger(self.name)
-
-    def __str__(self):
-        r = self.name
-        if getattr(self, 'executor', None):
-            r += f'({str(self.executor)})'
-        return r
-
-    def handle(self, msg: 'Message') -> 'BasePea':
-        """Register the current message to this pea, so that all message-related properties are up-to-date, including
-        :attr:`request`, :attr:`prev_requests`, :attr:`message`, :attr:`prev_messages`. And then call the executor to handle
-        this message if its envelope's  status is not ERROR, else skip handling of message.
+        self.daemon = args.daemon  #: required here to set process/thread daemon
 
-        :param msg: the message received
-        """
-
-        if self.expect_parts > 1 and self.expect_parts > len(self.partial_requests):
-            # NOTE: reduce priority is higher than chain exception
-            # otherwise a reducer will lose its function when eailier pods raise exception
-            raise NoExplicitMessage
+        self.name = self.args.name or self.__class__.__name__
+        self.is_ready = _get_event(self)
+        self.is_shutdown = _get_event(self)
+        self.ready_or_shutdown = _make_or_event(self, self.is_ready, self.is_shutdown)
+        self.logger = JinaLogger(self.name,
+                                 log_id=self.args.log_id,
+                                 log_config=self.args.log_config)
 
-        if msg.envelope.status.code != jina_pb2.StatusProto.ERROR or self.args.skip_on_error < SkipOnErrorType.HANDLE:
-            self.executor(self.request_type)
-        else:
-            raise ChainedPodException
-        return self
+        self._envs = {'JINA_POD_NAME': self.name,
+                      'JINA_LOG_ID': self.args.log_id}
+        if self.args.env:
+            self._envs.update(self.args.env)
 
-    @property
-    def is_idle(self) -> bool:
-        """Return ``True`` when current time is ``max_idle_time`` seconds late than the last active time"""
-        return (time.perf_counter() - self.last_active_time) > self.args.max_idle_time
+        try:
+            self.runtime = self._get_runtime_cls()(self.args)  # type: 'BaseRuntime'
+        except Exception as ex:
+            self.logger.error(f'{ex!r} during {self.runtime_cls.__init__!r}', exc_info=self.args.show_exc_info)
+            raise RuntimeFailToStart from ex
 
-    @property
-    def request(self) -> 'Request':
-        """Get the current request body inside the protobuf message"""
-        return self._request
+    def run(self):
```

Comment:
having a specializrd Pea to handle ZEQRuntime makes the Runtime of the main part of the logic much clearer and robust. instead of adapting so many different runtimes to be handled in the same exact way.

Reply:
one of the big headache is the explosive polymorphism on `Pod`, `Pea` and `Runtime`. In this PR I killed all subclasses of `Pod` and `Pea`, only `Runtime` allows polymorphism.